### PR TITLE
chore(vrl): Add `Kind::remove`

### PR DESCRIPTION
**Pull request by fuchsnj**
closes https://github.com/vectordotdev/vector/issues/13460
makes progress on https://github.com/vectordotdev/vector/issues/7070

This replaces the `Kind::remove_at_path` with a new `Kind::remove`.

- The previous implementation used the old lookup code, which has been migrated in the new implementation.
- Previously many deletions were not supported, such as coalescing, negative indexing, or compaction. All possible removals are now fully supported.

### CODE REVIEW

File: lib/value/src/kind/collection.rs [hunk 1]
```diff
@@ -266,6 +280,25 @@ impl<T: Ord + Clone> Collection<T> {
...
+}
+
+/// Collections have an "unknown" component, so it can't know in all cases if the value this
+/// collection represents is actually empty/not empty, so the state is represented with 3 variants.
+#[derive(Debug)]
+pub enum EmptyState {
+    // The collection is guaranteed to be empty.
+    Always,
+    // The collection may or may not actually be empty. There is not enough type information to
+    // determine. (There are unknown fields/indices that may exist, but there are no known values)
```
Comment (pablosichert) [inline lib/value/src/kind/collection.rs:296]:
```suggestion
    // The collection may or may not actually be empty. There is not enough type information to
    // determine. (There are unknown fields/indices that may exist, but there are no known values.)
```

File: lib/value/src/kind/collection/index.rs [hunk 2]
```diff
@@ -22,7 +23,35 @@ impl Collection<Index> {
...
+                if kind.contains_any_defined() {
+                    Some(i.to_usize())
+                } else {
+                    None
+                }
+            })
+            .max()
+    }
+
+    /// Converts a negative index to a positive index (only if the exact positive index is known)
```
Comment (pablosichert) [inline lib/value/src/kind/collection/index.rs:38]:
```suggestion
    /// Converts a negative index to a positive index (only if the exact positive index is known).
```

File: lib/value/src/kind/collection/index.rs [hunk 3]
```diff
@@ -41,6 +70,26 @@ impl Collection<Index> {
             Some(self.min_length())
         }
     }
+
+    /// Removes the known value at the given index and shifts the
+    /// elements to the left
```
Comment (pablosichert) [inline lib/value/src/kind/collection/index.rs:75]:
```suggestion
    /// Removes the known value at the given index and shifts the
    /// elements to the left.
```

File: lib/value/src/kind/comparison.rs [hunk 4]
```diff
@@ -399,6 +399,19 @@ impl Kind {
     pub const fn contains_object(&self) -> bool {
         self.object.is_some() || self.is_never()
     }
+
+    /// Returns `true` if the type contains _at least_ one non-collection type
```
Comment (pablosichert) [inline lib/value/src/kind/comparison.rs:403]:
```suggestion
    /// Returns `true` if the type contains _at least_ one non-collection type.
```

File: lib/value/src/kind/crud/remove.rs [hunk 5]
```diff
@@ -0,0 +1,922 @@
...
+
+    /// Combines two sets of options. Each setting is "or"ed together.
+    fn or(self, other: Self) -> Self {
+        Self::new(
+            self.should_compact() || other.should_compact(),
+            self.should_not_compact() || other.should_not_compact(),
+        )
+    }
+
+    /// If the value is true, the `should_compact` option is set to false
```
Comment (pablosichert) [inline lib/value/src/kind/crud/remove.rs:242]:
```suggestion
    /// If the value is true, the `should_compact` option is set to false.
```

**Review by pablosichert — COMMENTED** [review]
First round of pedantic suggestions 😄

File: lib/value/src/kind/collection/index.rs [hunk 6]
```diff
@@ -41,6 +70,26 @@ impl Collection<Index> {
...
+    pub fn remove_shift(&mut self, index: usize) {
+        let min_length = self.min_length();
+        self.known_mut().remove(&index.into());
+        for i in index..min_length {
```
Comment (pablosichert) [inline lib/value/src/kind/collection/index.rs:79]: What happens here if the collection has a gap in the indices, like `0`, `1`, `3`, `4`, and `1` is removed. Do we intentionally not shift the indices after the gap?

**Review by pablosichert — APPROVED** [review]
The `remove` logic is quite involved, I didn't try to understand every nuance of it. That being said, there was nothing that stood out as particularly odd to me.  +1 for the extensive set of tests to cover this area.

File: lib/value/src/kind/crud/remove.rs [hunk 7]
```diff
@@ -0,0 +1,922 @@
...
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum CompactOptions {
+    Always,
+    Maybe,
+    Never,
+}
```
Comment (JeanMertz) [inline lib/value/src/kind/crud/remove.rs:177]: It'd be good to add some inline comments here, even though this is an internal type.

**Review by JeanMertz — APPROVED** [review]
This looks great! Thanks for doing this.

**Pull request merged**